File: js-engine/src/ciphers/aead.ts
/* eslint-disable @typescript-eslint/no-namespace, no-inner-declarations */

import { K128, K192, K256 } from "../const";
import { CryptoError, ERROR_CODE } from "../@internals/errors";


abstract class AEADMode {
  public static for(alg: AEAD.ALGORITHM | AEAD.AlgorithmName): AEADMode {
    switch(alg) {
      case "aes-ccm":
      case "aes-gcm":
      case AEAD.ALGORITHM.AES_CCM:
      case AEAD.ALGORITHM.AES_GCM:
        return new AEAD.AESMode(alg === "aes-ccm" || alg === AEAD.ALGORITHM.AES_CCM ? "aes-ccm" : "aes-gcm");
      case "chacha20-poly1305":
      case AEAD.ALGORITHM.CHACHA20_POLY1305:
        return new AEAD.ChaCha20Mode();
      default:
        throw new CryptoError(`[AEADMode] Unknown or invalid algorithm ID "${alg}"`, ERROR_CODE.E_CRYPTO_INVALID_ARGUMENT);
    }
  }

  public abstract readonly basename: Lowercase<string>;
  public abstract readonly agid: number;
  public abstract readonly allowedKeySizeInBytes: readonly number[];
  public abstract readonly ivLength: number;
  public abstract readonly tagLength: number;
  public abstract readonly webTagLengthInBytes?: number;

  public getNodeName(ksize: number = K256): string {
    if(ksize < 100) {
      ksize *= 8;
    }

    if(this.basename.startsWith("aes-")) {
      const parts = this.basename.split("-");

      if(parts.length !== 2) {
        throw new CryptoError("[AEADMode] Invalid basename for AES cipher");
      }

      return `aes-${ksize}-${parts[1].trim()}`;
    }

    return this.basename;
  }

  public getWebCryptoName(): string {
    if([
      "aes-gcm",
      "aes-ccm",
      "chacha20-poly1305",
    ].includes(this.basename.toLowerCase()))
      return this.basename.toUpperCase();

    throw new CryptoError("[AEADMode] Invalid basename for AEAD cipher");
  }
}

export namespace AEAD {
  export type AlgorithmName = "aes-gcm" | "aes-ccm" | "chacha20-poly1305";

  export interface CipherMode extends AEADMode { }

  export const enum ALGORITHM {
    AES_GCM = 0x15,
    AES_CCM = 0x1C,
    CHACHA20_POLY1305 = 0xC5,
  }

  export class AESMode extends AEADMode {
    public override readonly basename: AlgorithmName;
    public override readonly agid: number;
    public override readonly allowedKeySizeInBytes: readonly number[];
    public override readonly ivLength: number;
    public override readonly tagLength: number;
    public override readonly webTagLengthInBytes: number;

    public constructor(mode: Exclude<AlgorithmName, "chacha20-poly1305"> = "aes-gcm") {
      if(!["aes-gcm", "aes-ccm"].includes(mode)) {
        throw new CryptoError(`[AESMode] Invalid algorithm name "${mode}"`, ERROR_CODE.E_CRYPTO_INVALID_ARGUMENT);
      }

      super();
      this.basename = mode;
      this.allowedKeySizeInBytes = Object.freeze([ K128, K192, K256 ]);

      if(mode === "aes-ccm") {
        this.agid = ALGORITHM.AES_CCM;
      } else {
        this.agid = ALGORITHM.AES_GCM;
      }

      this.ivLength = mode === "aes-ccm" ? 0xD : 0xC;
      this.tagLength = 0x10;
      this.webTagLengthInBytes = 0xC;
    }
  }

  export class ChaCha20Mode extends AEADMode {
    public override readonly basename: AlgorithmName;
    public override readonly agid: number;
    public override readonly allowedKeySizeInBytes: readonly number[];
    public override readonly ivLength: number;
    public override readonly tagLength: number;
    public override readonly webTagLengthInBytes: number;

    public constructor() {
      super();

      this.basename = "chacha20-poly1305";
      this.allowedKeySizeInBytes = Object.freeze([ K256 ]);
      this.agid = ALGORITHM.CHACHA20_POLY1305;
      this.ivLength = 0xC;
      this.tagLength = 0x10;
      this.webTagLengthInBytes = 0x10;
    }
  }

  export function isMode(arg: unknown): arg is CipherMode {
    return arg instanceof AEADMode;
  }
}

File: engine/main/ciphers/aead.c
#ifndef __CSUITE_CIPAEAD_MODULE__
#define __CSUITE_CIPAEAD_MODULE__ 0x1


#include "aead.h"


typedef enum {
  AES_GCM = 0x15,
  AES_CCM = 0x1C,
  CHACHA20_POLY1305 = 0xC5,
} AGID;

static const unsigned char AES_KEY_SIZES[] = { 0x10, 0x18, 0x20 };
#define AES_KEY_SIZES_COUNT (sizeof(AES_KEY_SIZES) / sizeof(unsigned char))

static const unsigned char CHACHA20_KEY_SIZES[] = { 0x20 };
#define CHACHA20_KEY_SIZES_COUNT (sizeof(CHACHA20_KEY_SIZES) / sizeof(unsigned char))


static AEADMode AEAD_AES_GCM_MODE = {
  .basename = "aes-gcm",
  .agid = AES_GCM,
  .iv_length = 0xC,
  .tag_length = 0x10,
  .allowed_key_sizes = AES_KEY_SIZES,
  .allowed_key_sizes_count = AES_KEY_SIZES_COUNT
};

static AEADMode AEAD_AES_CCM_MODE = {
  .basename = "aes-ccm",
  .agid = AES_CCM,
  .iv_length = 0xD,
  .tag_length = 0x10,
  .allowed_key_sizes = AES_KEY_SIZES,
  .allowed_key_sizes_count = AES_KEY_SIZES_COUNT
};

static AEADMode AEAD_CHACHA20_MODE = {
  .basename = "chacha20-poly1305",
  .agid = CHACHA20_POLY1305,
  .iv_length = 0xC,
  .tag_length = 0x10,
  .allowed_key_sizes = CHACHA20_KEY_SIZES,
  .allowed_key_sizes_count = CHACHA20_KEY_SIZES_COUNT
};


AEADMode* aead_aes(const char *mode) {
  if(!mode) return NULL;

  if(strcmp(mode, "ccm") == 0 || strcmp(mode, "CCM") == 0)
    return &AEAD_AES_CCM_MODE;

  if(strcmp(mode, "gcm") == 0 || strcmp(mode, "GCM") == 0)
    return &AEAD_AES_GCM_MODE;

  return NULL;
}

AEADMode* aead_chacha20() {
  return &AEAD_CHACHA20_MODE;
}

#endif
